### ISSUE TITLE
Crash log on HandleForceSpeedChangeAck

### ISSUE THREAD

**Issue by ckegg** [TrinityCore/TrinityCore#26642]
**Description:**
Not sure if it is related to https://github.com/TrinityCore/TrinityCore/pull/26561
worldserver.conf : AntiCheat.PendingMoveChangesTimeoutTime = 4000

[crash-2021-06-27-21-09-00.log](https://github.com/TrinityCore/TrinityCore/files/6722560/crash-2021-06-27-21-09-00.log)
[crash-2021-06-27-23-11-33.log](https://github.com/TrinityCore/TrinityCore/files/6722561/crash-2021-06-27-23-11-33.log)
[crash-2021-06-27-23-58-04.log](https://github.com/TrinityCore/TrinityCore/files/6722562/crash-2021-06-27-23-58-04.log)
[crash-2021-06-28-05-00-45.log](https://github.com/TrinityCore/TrinityCore/files/6722563/crash-2021-06-28-05-00-45.log)

**Expected behaviour:**
No crash

**Steps to reproduce the problem:**
Tried to reproduce but was unable to. Tried levitate, slow fall or whatever but dunno how crashes happened. Maybe lag?

**Branch(es):**
3.3.5

**TC rev. hash/commit:** 
c8ca48823d45dadb042a80637eff954483dd5328

**Operating system:** 
debian 11

**Comment by jackpoz** [comment]
@chaodhib could you please check this ?

**Comment by jackpoz** [comment]
```c++
REMOVE_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_FALLING_SLOW) && !mover->HasAuraType(SPELL_AURA_FEATHER_FALL),
        MOVEMENTFLAG_FALLING_SLOW);
```

`mover` is null at that point

**Comment by chaodhib** [comment]
@jackpoz will do

**Comment by ckegg** [comment]
Crash log with crashreport.gdb script 1.4
3.3.5 b596ea298a7b
[crash-2021-07-01-10-58-17.log](https://github.com/TrinityCore/TrinityCore/files/6745134/crash-2021-07-01-10-58-17.log)

**Comment by Aokromes** [comment]
maybe this is druids or shamans with knock back spells?

**Comment by ckegg** [comment]
>
> [crash-2021-07-01-10-58-17.log](https://github.com/TrinityCore/TrinityCore/files/6745134/crash-2021-07-01-10-58-17.log)

I dig deeper on all guids and they all were hunter characters.
Out of 5 logs, 3 crashed at this pos:  .go xyz 3547 299 121 571

**Comment by jackpoz** [comment]
@ckegg thanks for the crashlogs

**Comment by Aokromes** [comment]
ok, on that point you can take a drake to make one quest, maybe it's needed to use water walk or feather fall and take taxi?

**Comment by jackpoz** [comment]
https://github.com/TrinityCore/TrinityCore/commit/fdbff48808ee65480bc7915c55d154e6cb46b00e should fix this

**Comment by jackpoz** [comment]
> worldserver.conf : AntiCheat.PendingMoveChangesTimeoutTime = 4000

@ckegg could you set this to 0 for a few weeks just to test only the rest of the code ?

**Comment by jackpoz** [comment]
well, it seems @Aokromes reproduced again the issue, so not fixed with the PR. @chaodhib 
#0 0x427ab81 in Trinity::Assert(char const*, int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char const*, char const*, ...) /trinity/trinity/sources/src/common/Debugging/Errors.cpp:90:5
    #1 0x3579a94 in WorldSession::ReadMovementInfo(WorldPacket&, MovementInfo*) /trinity/trinity/sources/src/server/game/Server/WorldSession.cpp:931:5

**Comment by Aokromes** [comment]
make rogue /* maybe with sprint glyph */
.tele wyrmrest
use cloak of shadows
talk with the drake who fly to middle or bottom while cloak is enabled and take taxi.

**Comment by Aokromes** [comment]
@chaodhib

**Comment by chaodhib** [comment]
Thank you. I'll look into it tonight.

**Comment by jackpoz** [comment]
https://github.com/TrinityCore/TrinityCore/commit/0c90970b0668df93356d035ecff8124b5dc5c44d tries to avoid asserting 24/7

**Comment by chaodhib** [comment]
The line `Unit* mover = GetGameClient()->GetActivelyMovedUnit();` was actually a mistake. The client can ACK on units that are not the currently "actively moved unit". https://github.com/TrinityCore/TrinityCore/commit/fdbff48808ee65480bc7915c55d154e6cb46b00e fixes that. Better keep it this way. 

The problem here is the guid in MovementInfo. Some handler don't set it before calling `WorldSession::ReadMovementInfo`. So `mi->guid == 0x0` and `ObjectAccessor::GetUnit(*GetPlayer(), mi->guid)` is `null`.

I'll submit a PR shortly and rollback https://github.com/TrinityCore/TrinityCore/commit/0c90970b0668df93356d035ecff8124b5dc5c44d.

**Comment by jackpoz** [comment]
Leave the check for IsEmpty() and the return when it's empty.

**Comment by Shauren** [comment]
ACK should never happen on different units, 3 possible reasons here
1) Server bug - sending opcode that needs ACK to incorrect unit (for example using SendMessageToSet instead of SendDirectMessage)
2) Insufficient movement packet sequencing reversing
3) Cheaters

**Comment by chaodhib** [comment]
@jackpoz 

> Leave the check for IsEmpty() and the return when it's empty.

Sorry, which `IsEmpty()` are you talking about?

> If the client can ack on different units, then we don't have to assert at all ? Just return if mover is null ?

if `client->IsAllowedToMove(guid)` or `client->IsAllowedToMove(guid)` was called before `ReadMovementInfo`, mover should never be null in the first place. That means the server is allow the client to move a unit that doesn't exist anymore. if that's the case, there is a bug in the code and an ASSERT is definitively warranted imo.

**Comment by chaodhib** [comment]
@Shauren the client can ACK on a unit that is not the active mover. See the [comment here](https://github.com/TrinityCore/TrinityCore/blob/0c90970b0668df93356d035ecff8124b5dc5c44d/src/server/game/Handlers/MovementHandler.cpp#L459).

**Comment by jackpoz** [comment]
https://github.com/TrinityCore/TrinityCore/commit/0c90970b0668df93356d035ecff8124b5dc5c44d#diff-d8eeab088d8241a103987b65f23a0b9f5f008df2b379fe9bb3fc9e176b777a9eR933 there's only one IsEmpty() in that commit

**Comment by jackpoz** [comment]
Also let's stop using assert on every bug that exist in tc, we have so many different log levels, let's use them for once and allow players to play and don't get a crash every 5 minutes.

There are so many asserts that are just a sign of lazy developers not wanting to add a if check. Is it an error ? Use TC_LOG_ERROR

**Comment by chaodhib** [comment]
The issue should fixed by https://github.com/TrinityCore/TrinityCore/commit/7314d5b9925ba16a17f1f575d96e7bce5947facd.

@jackpoz ReadMovementInfo should always be called after client->IsAllowedToMove(guid) or client->IsAllowedToMove(guid) but sure i'll add it

**Comment by chaodhib** [comment]
> 
> 
> Also let's stop using assert on every bug that exist in tc, we have so many different log levels, let's use them for once and allow players to play and don't get a crash every 5 minutes.
> 
> There are so many asserts that are just a sign of lazy developers not wanting to add a if check. Is it an error ? Use TC_LOG_ERROR

That makes sense. Done in https://github.com/TrinityCore/TrinityCore/commit/4806b3a8602ad15e48130bb9b49d020d3beb90e4.

Let's close this issue and reopen if the problem persists.

Closed by 7314d5b.
